Lib/zipfile/_path.py

```diff
@@ -254,12 +254,12 @@ def open(self, mode='r', *args, pwd=None, **kwargs):
             raise FileNotFoundError(self)
         stream = self.root.open(self.at, zip_mode, pwd=pwd)
         if 'b' in mode:
-            if args or kwargs:
+            if encoding is not None or args or kwargs:
                 raise ValueError("encoding args invalid for binary operation")
             return stream
         else:
-            kwargs["encoding"] = io.text_encoding(kwargs.get("encoding"))
-        return io.TextIOWrapper(stream, *args, **kwargs)
+            encoding = io.text_encoding(encoding)
+        return io.TextIOWrapper(stream, *args, encoding=encoding, **kwargs)
```

Comment:
(reposted as it seems to have been lost in my original submission).
I believe this change is incorrect, as it has re-ordered the positional args. If one were to pass a positional arg, it would cause an error because 'encoding' would be passed both positionally and by keyword. Same for `read_text`.

Reply:
indeed. good catch! I've reworked how this goes which keeps the complication in a single spot here rather than scattered. PTAL.